module/extension/Director/tests/wait_task_cancel.cpp

```diff
@@ -0,0 +1,137 @@
+/*
+ * MIT License
+ *
+ * Copyright (c) 2022 NUbots
+ *
+ * This file is part of the NUbots codebase.
+ * See https://github.com/NUbots/NUbots for further info.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#include <catch2/catch_test_macros.hpp>
+#include <nuclear>
+
+#include "Director.hpp"
+#include "TestBase.hpp"
+#include "util/diff_string.hpp"
+
+// Anonymous namespace to avoid name collisions
+namespace {
+
+    struct SimpleTask {
+        SimpleTask(const bool& wait_) : wait(wait_) {}
+        bool wait = false;
+    };
+
+    std::vector<std::string> events;
+
+    class TestReactor : public TestBase<TestReactor, 5> {
+    public:
+        explicit TestReactor(std::unique_ptr<NUClear::Environment> environment) : TestBase(std::move(environment)) {
+            // Wait does not want to shut down automatically at the end of the steps, to allow the chrono task to finish
+            auto_shutdown = false;
+
+            on<Provide<SimpleTask>>().then([this](const SimpleTask& task, const RunReason& run_reason) {
+                if (run_reason == RunReason::SUBTASK_DONE) {
+                    events.push_back("task executed, done waiting");
+                    powerplant.shutdown();
+                }
+                else if (task.wait) {
+                    // Rerun the provider after 100ms with Wait
+                    events.push_back("task executed, waiting");
+                    emit<Task>(std::make_unique<Wait>(NUClear::clock::now() + std::chrono::milliseconds(100)));
+                }
+                else {
+                    events.push_back("task executed, no wait");
+                }
+            });
+
+            /**************
+             * TEST STEPS *
+             **************/
+            on<Trigger<Step<1>>, Priority::LOW>().then([this] {
+                // Freeze time
+                emit(std::make_unique<NUClear::message::TimeTravel>(NUClear::clock::now(),
```

Comment:
Maybe time travel to absolute times in this test and then include now in the log lines. Then you can test that it's waiting not just running in order